enhancements/cloud-integration/tokenized-auth-enablement-operators-on-cloud.md

```diff
@@ -0,0 +1,503 @@
+API field on the CredentialsRequest spec. When CCO acquires a CredentialsRequest it will mint a Secret (as today 
+mostly) with new functionality: Adding to the Secret a path to the projected ServiceAccount token (static for OCP) 
+and a roleARN (from the new field on the CredentialsRequest).
+
+Validation of CredentialsRequest by Token mode CCO? Maybe. Currently, CCO gets some permissions under ROSA (unused),
+could expand these permissions to include tools like AWS's Policy Simulator such that it could validate a role has 
+permissions, and with this CCO could be the alerting mechanism for a changed CredentialsRequest without sufficient 
+permissions.
+
+**Pod Identity Webhook mode** (Deployed on classic ROSA, ROSA on HyperShift, standard OCP and Hypershift)
+
+Will work by annotating the ServiceAccount triggering the projection of the service account tokens into pods created for
+the operator (same method and resultant actions as: https://github.com/openshift/hypershift/pull/1351) the README here 
+gives an AWS specific example: https://github.com/openshift/aws-pod-identity-webhook)
+
+Possibly some changes to logic for Pod Identity Webhooks as needed for fine-grained credential
+management to allow for multi-tenancy (see definition earlier). Also, need to add pod identity webhooks targeting other
+cloud providers and change the annotations to more generic naming.
+
+**HyperShift changes**: Include Cloud Credential Operator with "Token" mode (see above). Allows for processing of 
+CredentialsRequest objects added by operators.
+
+**OperatorHub and Console changes**: Allow for import of additional, well-known, bundle annotations which will result 
+in ENV variables on the Subscription object. Show operator is enabled for Token-based use based on Console determining 
+cluster is running in token-based authentication enabled cluster by reading the `.spec.serviceAccountIssuer` from the Authentication CR, `.spec.platformSpec.type` from the Infrastructure CR,`.spec.credentialsMode` from the CloudCredentials CR  AND
+bundle has token-based auth enabled annotation. This info should be written to a new field on Subscription objects.
+The Subscription fields will allow UX for the information needed by CCO or webhook, for input of the cloud credentials. 
+Generate a manual ack for operator upgrade when cloud resources are involved. This can be determined by parsing the 
+Subscription objects. Ensures admin signs off that cloud resources are provisioned and working as intended before an 
+operator upgrade.
+
+**Operator team/ Operator SDK changes**: Follow new guidelines for allowing for the operator to work on token auth 
+enabled cluster. New guidelines would include the following to use CCO Token mode:
+
+- changing CRD to add CredentialRequest references and putting those referenced
+- CredentialRequests into a defined directory in the bundle
```

Comment:
So "extend olm to support CredRequest resource types" becomes a requirement, right?

(as long as the solution is going through CCO and not doing something else like directly creating secrets, or using the aws pod identity webhook)

Reply:
So the idea is that operators that talk to cloud APIs can embed a static CredentialsRequest in their bundle, and that is expected to work on all potential clusters and cluster configurations where anyone would want to install that operator?

Operators don't need to probe that the CredentialRequest API is available and to see whether STS mode is enabled or not to decide whether to create and what to put in the CredentialRequest?

Do any of these operators ever make sense to deploy on a cluster where CCO is not running?
Do bundles that include CredentialsRequest need to declare a GVK dependency on that API?

Reply:
In the case where it's the operand that needs a CredReq, OLM doesn't need to know about the CredReq, the operator would apply it with the operand deployment / setup. I agree we want it in a consistent place but I'm not sure the bundle manifests is the right spot because it's not something that we know we want applied at install time.

Reply:
At install time we know we need to give the operator the role ARN for example for STS. But that's it, it will decide when to create the CredReq.

Reply:
OLM does not support CredReqs in the bundle. It _does_ support some Console APIs in the bundle, so there's at least precedent for having OCP-specific APIs supported this way.

It isn't there today, and would need to be added if we wanted OLM to be able to create CredReqs at install time.

Reply:
Going to resolve this thread since OLM creating CredentialsRequests is not in scope for this enhancement. OLM may be later extended to support CredentialsRequests but that would be handled in a separate enhancement.

Reply:
okay, it's been a week, marking resolved, James' reasoning is sound on reason to resolve.